### PR TITLE
IGNITE-17941 Add learners to Raft API

### PR DESCRIPTION
**Pull request by sashapolo**
https://issues.apache.org/jira/browse/IGNITE-17941

### CODE REVIEW

File: modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+    @Test
+    public void testLostLearners() throws Exception {
+        RaftNode follower = nodes.get(0);
+        List<RaftNode> learners = nodes.subList(1, nodes.size());
+
+        List<CompletableFuture<RaftGroupService>> services = nodes.stream()
+                .map(node -> startRaftGroup(node, new TestRaftGroupListener(), List.of(follower), learners))
+                .collect(toList());
+
+        // Wait for the leader to be elected.
```
Comment (ibessonov) [inline modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java:253]: Comment got duplicated

File: modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+    }
+
+    private CompletableFuture<RaftGroupService> startRaftGroup(
+            RaftNode raftNode,
+            RaftGroupListener listener,
+            List<RaftNode> peers,
+            List<RaftNode> learners
+    ) {
+        try {
+            CompletableFuture<RaftGroupService> future = raftNode.loza.prepareRaftGroup(
```
Comment (ibessonov) [inline modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java:272]: There should be a version of this method that takes listeners directly, not their suppliers. Or maybe I'm wrong

File: modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+                    RAFT_GROUP_NAME,
+                    peers.stream().map(RaftNode::localMember).collect(toList()),
+                    learners.stream().map(RaftNode::localMember).collect(toList()),
+                    () -> listener,
+                    () -> noopLsnr,
+                    RaftGroupOptions.defaults()
+            );
+
+            return future.thenApply(s -> {
+                s.timeout(100);
```
Comment (ibessonov) [inline modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java:282]: Please add a comment for this line. I believe it is here to make tests faster

File: modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+
+                return s;
+            });
+        } catch (NodeStoppingException e) {
+            throw new RuntimeException(e);
```
Comment (ibessonov) [inline modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java:287]: Why not `fail(...)`?

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java [hunk 5]
```diff
@@ -305,25 +280,58 @@ public CompletableFuture<RaftGroupService> startRaftGroupService(
...
+            String grpId,
+            List<Peer> peers,
+            List<Peer> learners,
+            RaftGroupListener lsnr,
+            RaftGroupEventsListener raftGrpEvtsLsnr,
+            RaftGroupOptions groupOptions
+    ) {
+        assert !peers.isEmpty();
+
+        LOG.info("Start new raft node for group={} with initial peers={}", grpId, peers);
```
Comment (ibessonov) [inline modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java:300]: Should we put it into `if (infoEnabled())`? That's usually a good practice

File: modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+        public boolean onSnapshotLoad(Path path) {
+            return true;
+        }
+
+        @Override
+        public void onShutdown() {
+        }
+    }
+
+    private static class TestWriteCommand implements WriteCommand {
```
Comment (ibessonov) [inline modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java:327]: You know what, can you make it a network message in advance? We're porting all commands right now, so you'll save us some time

File: modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/RaftGroupService.java [hunk 7]
```diff
@@ -239,18 +237,6 @@ public interface RaftGroupService {
...
-     * Runs a read command on a given peer.
-     *
-     * <p>Read commands can see stale data (in the past).
-     *
-     * @param peer Peer id.
-     * @param cmd  The command.
-     * @param <R>  Execution result type.
-     * @return A future with the execution result.
-     */
-    <R> CompletableFuture<R> run(Peer peer, ReadCommand cmd);
```
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/RaftGroupService.java:252]: This API is not used and is not intended to be used, NetworkMessages will be sent directly to the replica nodes

File: modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java (hunk 2 above)
Comment (sashapolo) [reply modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java:272]: Why?

File: modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java (hunk 3 above)
Comment (sashapolo) [reply modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java:282]: Yes, you are right

File: modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java (hunk 4 above)
Comment (sashapolo) [reply modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java:287]: Java compiler does not know that `fail` always throws, so I will have to add `return null`, which is ugly

File: modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/RaftGroupService.java [hunk 8]
```diff
@@ -49,7 +49,7 @@ public interface RaftGroupService {
     /**
      * Returns group id.
      */
-    String groupId();
+    ReplicationGroupId groupId();
```
Comment (ibessonov) [inline modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/RaftGroupService.java:52]: Is this OK to make this replacement if we don't fully own a RAFT code? Can you consult with @alievmirza?

File: modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java [hunk 9]
```diff
@@ -73,19 +77,38 @@
  */
 @ExtendWith(ConfigurationExtension.class)
 public class ItLearnersTest extends IgniteAbstractTest {
-    private static final String RAFT_GROUP_NAME = "test";
+    private static final ReplicationGroupId RAFT_GROUP_ID = new ReplicationGroupId() {
+        @Override
+        public String toString() {
+            return "test";
```
Comment (ibessonov) [inline modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java:83]: Shouldn't there be a dedicated method for the id? `toString()` usually has a different meaning

File: modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java [hunk 10]
```diff
@@ -73,19 +77,38 @@
...
+        }
+    };
 
     private static final List<NetworkAddress> ADDRS = List.of(
             new NetworkAddress("localhost", 5000),
             new NetworkAddress("localhost", 5001),
             new NetworkAddress("localhost", 5002)
     );
 
+    private static final TestRaftMessagesFactory msgFactory = new TestRaftMessagesFactory();
```
Comment (ibessonov) [inline modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java:93]: Constants should be named in upper case

File: modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java (hunk 2 above)
Comment (ibessonov) [reply modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java:296]: I mistaken it for `start...` method. Anyway, this approach with suppliers of listeners does stink, if you ask me, I tried to partially eliminate it

File: modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/RaftGroupService.java (hunk 8 above)
Comment (ibessonov) [reply modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/RaftGroupService.java:52]: I guess it's OK after all, I'm just paranoid about raft packages

Comment (sashapolo) [reply modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/RaftGroupService.java:52]: What replacement are you talking about?

File: modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java (hunk 9 above)
Comment (sashapolo) [reply modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java:83]: Again, I don't understand. This class wasn't introduced by me, I'm just playing by the rules introduced by others

File: modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java (hunk 10 above)
Comment (sashapolo) [reply modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java:93]: Technically this is not a constant, it's a static factory)

File: modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/RaftGroupService.java (hunk 8 above)
Comment (ibessonov) [reply modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/RaftGroupService.java:52]: String -> ReplicationGroupId

File: modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java (hunk 9 above)
Comment (ibessonov) [reply modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java:83]: Didn't noticed that at first

File: modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java (hunk 10 above)
Comment (ibessonov) [reply modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java:93]: You know what I mean)

File: modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java (hunk 4 above)
Comment (ibessonov) [reply modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java:312]: I see

File: modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/RaftGroupService.java (hunk 8 above)
Comment (alievmirza) [reply modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/RaftGroupService.java:52]: Seems that this changes was introduced by Vlad today, I'm not aware of them

**Review by ibessonov — APPROVED** [review]

**Pull request merged**